### ISSUE TITLE
Batchnorm1d with affine=False slower than affine=True

### ISSUE THREAD

**Issue by samuelbroscheit** [pytorch/pytorch#5374]
- OS: Linux 3.16.0-4-amd64 #1 SMP Debian 3.16.43-2+deb8u5 (2017-09-19) x86_64 GNU/Linux
- PyTorch version: 3.0
- How you installed PyTorch (conda, pip, source): pip 
- Python version: 3.6
- CUDA/cuDNN version: '8.0.61'
- GPU models and configuration: GeForce GTX 1080Ti
- GCC version (if compiling from source): -

```
import torch

hid_size = 4096
batch_size = 100
epochs = 1000

net = torch.nn.Sequential(
    torch.nn.Linear(hid_size,hid_size),
    torch.nn.BatchNorm1d(affine=False, num_features=hid_size),
    torch.nn.Linear(hid_size,2),
)
net.cuda()

x = torch.autograd.Variable(torch.randn(batch_size,hid_size))
y = torch.autograd.Variable(torch.LongTensor([1, 0] * 5))

x = x.cuda()
y = y.cuda()

loss = torch.nn.CrossEntropyLoss()
optim = torch.optim.SGD(net.parameters(), lr=0.1)

def train():
    for i in range(epochs):
        optim.zero_grad()
        p = net(x)
        loss(p, y).backward()
        optim.step()
```
%timeit train()

GPU affine=False 1.98 s ± 860 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
GPU affine=True 1.93 s ± 1.58 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

I was expecting a speedup, but in my model "affine=False" even led to a 15% decrease in throughput. Is this expected?

**Comment by SsnL** [comment]
Yes, it is expected. With affine, the cudnn path is used. Without, we use the thnn path.

**Comment by samuelbroscheit** [comment]
I see, thanks !

**Comment by SsnL** [comment]
By the way, cuda kernels are run asynchronously. So, when benchmarking cuda code, you should use `torch.cuda.synchronize()` to get the correct timings. e.g., 
```
t = 0
for i in range(epochs):
    torch.cuda.synchronize()
    t0 = time.time()
    optim.zero_grad()
    p = net(x)
    loss(p, y).backward()
    optim.step()
    torch.cuda.synchronize()
    t += time.time() - t0
```

**Comment by samuelbroscheit** [comment]
The script below has two networks Net 1 and Net 2. The difference between Net 1 and Net 2 is

Net 1 (standard Batchnorm1d) 

    if self.affine:
        self.weight = Parameter(torch.Tensor(num_features))
        self.bias = Parameter(torch.Tensor(num_features))
    else:
        self.register_parameter('weight', None)
        self.register_parameter('bias', None)
           

Net 2 where the init of Batchnorm is changed to

    if self.affine:
        self.weight = Parameter(torch.Tensor(num_features))
        self.bias = Parameter(torch.Tensor(num_features))
    else:
        self.weight = Parameter(torch.ones(num_features), requires_grad=False)
        self.bias = Parameter(torch.zeros(num_features), requires_grad=False)

I ran the script below with pytorch 3.1 and pytorch 4.1: 

pytorch 3.1

    hid_sz	Net 1	Net 2
        32	1.03	1.43
        64	0.94	0.87
       128	0.91	0.90
       256	0.88	0.80
       512	1.01	0.94
      1024	1.93	0.96
      2048	3.52	1.54
      4096	8.05	4.72

pytorch 4.1

    hid_sz	Net 1	Net 2
        32	1.12	1.21
        64	1.14	1.22
       128	1.15	1.22
       256	1.15	1.23
       512	1.29	1.29
      1024	2.04	1.32
      2048	3.65	1.75
      4096	8.42	4.77

The timings show that from a hidden size of 1024 onwards it is faster to use the Batchnorm from Net 2, i.e. to do a identity transformation in CUDNN instead of doing THNN BN. Also there seems to be a regression from 3.1 to 4.1. In general, for smaller hidden sizes THNN is not significantly faster.

The simple fix would be to implement the init for _Batchnorm like in the example code, i.e. do a identity transform in CUDNN. 

## To Reproduce

Steps to reproduce the behavior:



    import time
    import torch
    from torch.nn import Parameter
    from torch.nn.modules.batchnorm import _BatchNorm
    
    class BatchNorm1d(_BatchNorm):
    
        def __init__(self, num_features, eps=1e-5, momentum=0.1, affine=True):
            super(BatchNorm1d, self).__init__(num_features, eps=eps, momentum=momentum, affine=affine)
            self.num_features = num_features
            self.affine = affine
            self.eps = eps
            self.momentum = momentum
            if self.affine:
                self.weight = Parameter(torch.Tensor(num_features))
                self.bias = Parameter(torch.Tensor(num_features))
            else:
                self.weight = Parameter(torch.ones(num_features), requires_grad=False)
                self.bias = Parameter(torch.zeros(num_features), requires_grad=False)
            self.register_buffer('running_mean', torch.zeros(num_features))
            self.register_buffer('running_var', torch.ones(num_features))
            self.reset_parameters()
    
    
    batch_size = 500
    epochs = 1000
    
    for hid_size in [32*2**i for i in range(8)]:
    
        net_1 = torch.nn.Sequential(
            torch.nn.Linear(hid_size,hid_size),
            torch.nn.BatchNorm1d(affine=False, num_features=hid_size),
            torch.nn.Linear(hid_size,2),
        )
        net_1.cuda()
    
        net_2 = torch.nn.Sequential(
            torch.nn.Linear(hid_size,hid_size),
            BatchNorm1d(affine=False, num_features=hid_size),
            torch.nn.Linear(hid_size,2),
        )
        net_2.cuda()
    
        x = torch.autograd.Variable(torch.randn(batch_size,hid_size))
        y = torch.autograd.Variable(torch.LongTensor([1, 0] * (batch_size//2)))
    
        x = x.cuda()
        y = y.cuda()
    
        for id, net in enumerate([net_1, net_2]):
    
            loss = torch.nn.CrossEntropyLoss()
            optim = torch.optim.SGD([p for p in net.parameters() if p.requires_grad == True], lr=0.1)
    
            t = 0
            for i in range(epochs):
                torch.cuda.synchronize()
                t0 = time.time()
                optim.zero_grad()
                p = net(x)
                loss(p, y).backward()
                optim.step()
                torch.cuda.synchronize()
                t += time.time() - t0
    
            print('Net {} with hidden size {}: {}'.format(id+1, hid_size, t))
